fuse_viz/src/relative_pose_2d_stamped_constraint_visual.cpp

```diff
@@ -137,6 +145,45 @@ void RelativePose2DStampedConstraintVisual::setConstraint(
   error_line_->addPoint(absolute_position_ogre);
   error_line_->addPoint(toOgre(pose2.getOrigin()));
 
+  // Update constraint loss error line:
+  loss_error_line_->clear();
+
+  auto loss_function = constraint.lossFunction();
+  if (loss_function)
+  {
+    // Evaluate cost function without loss:
+    const double position1[] = { pose1.getOrigin().getX(), pose1.getOrigin().getY() };
+    const double yaw1[] = { tf2::getYaw(pose1.getRotation()) };
+    const double position2[] = { pose2.getOrigin().getX(), pose2.getOrigin().getY() };
+    const double yaw2[] = { tf2::getYaw(pose2.getRotation()) };
+
+    const double* parameters[] = { position1, yaw1, position2, yaw2 };
+
+    auto cost_function = constraint.costFunction();
+
+    fuse_core::VectorXd residuals(cost_function->num_residuals());
+
+    cost_function->Evaluate(parameters, residuals.data(), nullptr);
+    delete cost_function;
+
+    const auto cost = residuals.squaredNorm();
+
+    // Evaluate cost with with loss:
```

Comment:
I'm not sure if the cost with loss is exactly `loss_cost[0]` as returned by `loss_function->Evaluate`. According to https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h#L99 I might be `0.5 * loss_cost[0]` instead.

I'm not sure because the implementations of many loss functions don't match the corresponding `\rho(x)` equations in http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24). Remember that in Ceres solver that argument pass is `s = x^2`. I've been comparing the `Tukey` loss function, because others have some subtleties like the `Huber` loss function being the `PseudoHuber` actually, among others. If you look at the `Trivial` one, which is `L2`, it actually looks like the `0.5 *` factor is applied after and I should apply it. :thinking:

Reply:
Both the cost with and without loss is multiplied by `0.5`: https://github.com/ceres-solver/ceres-solver/blob/c8e35e19fd3409bfbb00e5247300af381223e259/internal/ceres/residual_block.cc#L158-L165

So the implementation was correct. I've added some comments to make that clear.